File: transports/http-sse/src/codec.ts
// Pure SSE framing for GIK messages. No sockets here, so this is unit-testable on its
// own: encode a message to an `data: ...\n\n` frame, and parse a byte stream (which may
// split frames across chunks) back into messages.

import type { GIKMessage } from "@gik/kernel";

/** Encode one GIK message as a single SSE `data:` frame. */
export function encodeSseFrame(message: GIKMessage): string {
  return `data: ${JSON.stringify(message)}\n\n`;
}

/**
 * Incremental SSE frame parser. Feed it decoded chunks (which may contain partial or
 * multiple frames); it returns the GIK messages completed by each chunk. Non-`data:`
 * lines (comments/heartbeats like `:keep-alive`, `id:`, `event:`) are ignored per the
 * SSE spec.
 */
export class SseFrameParser {
  private buffer = "";

  push(chunk: string): GIKMessage[] {
    this.buffer += chunk;
    const messages: GIKMessage[] = [];
    let sep: number;
    while ((sep = this.buffer.indexOf("\n\n")) !== -1) {
      const frame = this.buffer.slice(0, sep);
      this.buffer = this.buffer.slice(sep + 2);
      const data = frame
        .split("\n")
        .filter((line) => line.startsWith("data:"))
        .map((line) => line.slice(5).replace(/^ /, ""))
        .join("\n");
      if (data) messages.push(JSON.parse(data) as GIKMessage);
    }
    return messages;
  }
}

File: adapters/dotnet/GenUI.Transport.HttpSse/Codec.cs
// Pure SSE framing for GUP messages — no sockets here, so it is unit-testable on its own:
// encode a message to a `data: ...\n\n` frame, and parse a byte stream (which may split frames
// across chunks) back into messages. The C# mirror of transports/http-sse/src/codec.ts.

using System.Text.Json.Nodes;

namespace GenUI.Transport.HttpSse;

public static class Codec
{
    /// <summary>Encode one enveloped GUP message as a single SSE <c>data:</c> frame.</summary>
    public static string EncodeSseFrame(JsonObject message) => $"data: {message.ToJsonString()}\n\n";
}

/// <summary>
/// Incremental SSE frame parser. Feed it decoded chunks (which may contain partial or multiple
/// frames); it returns the GUP messages completed by each chunk. Non-<c>data:</c> lines
/// (comments/heartbeats like <c>:keep-alive</c>, <c>id:</c>, <c>event:</c>) are ignored per the SSE spec.
/// </summary>
public sealed class SseFrameParser
{
    private string _buffer = string.Empty;

    public IReadOnlyList<JsonObject> Push(string chunk)
    {
        _buffer += chunk;
        var messages = new List<JsonObject>();
        int sep;
        while ((sep = _buffer.IndexOf("\n\n", StringComparison.Ordinal)) != -1)
        {
            var frame = _buffer[..sep];
            _buffer = _buffer[(sep + 2)..];

            var data = string.Join("\n", frame
                .Split('\n')
                .Where(line => line.StartsWith("data:", StringComparison.Ordinal))
                .Select(line =>
                {
                    var value = line[5..];
                    return value.StartsWith(' ') ? value[1..] : value;
                }));

            if (data.Length > 0 && JsonNode.Parse(data) is JsonObject obj)
                messages.Add(obj);
        }
        return messages;
    }
}
